app/controllers/api/v1/users/omniauth_callbacks_controller.rb

```diff
@@ -28,10 +30,24 @@ def failure
 
         private
 
+        def client_application
+          state = JSON.parse(request.env['action_dispatch.request.parameters']['state'])
+          app_client_id = state['client_id']
+
+          Doorkeeper::Application.find_by(uid: app_client_id)
+        end
+
         def render_success
-          # TODO: Render Doorkeeper Token
+          user = @user
```

Comment:
This looks a little bit od :)

What about using a private attribute reader instead?

```
private

attr_reader :user
```

Then you can call `user` in any instance method of this controller.

```suggestion
```

... and we do not need this line anymore :)